Sources/NIOHTTP1/NIOHTTPObjectAggregator.swift

```diff
@@ -0,0 +1,400 @@
+///
+/// This is achieved by buffering the contents of all received `HTTPServerRequestPart`
+/// messages until `HTTPServerRequestPart.end` is received, then assembling the
+/// full message and firing a channel read upstream with it. It is useful for when you do not
+/// want to deal with chunked messages and just want to receive everything at once, and
+/// are happy with the additional memory used and delay handling of the message until
+/// everything has been received.
+///
+/// `NIOHTTPServerRequestAggregator` may end up sending a `HTTPResponseHead`:
+/// - Response status `413 Request Entity Too Large` when either the
+///     `content-length` or the bytes received so far exceed `maxContentLength`.
+///
+/// `NIOHTTPServerRequestAggregator` may close the connection if it is impossible
+/// to recover:
+/// - If `content-length` is too large and `keep-alive` is off.
+/// - If the bytes received exceed `maxContentLength` and the client didn't signal
+///     `content-length`
+public final class NIOHTTPServerRequestAggregator: ChannelInboundHandler, RemovableChannelHandler {
+    public typealias InboundIn = HTTPServerRequestPart
+    public typealias InboundOut = NIOHTTPServerRequestFull
+
+    // Aggregator may generate responses of its own
+    public typealias OutboundOut = HTTPServerResponsePart
+
+    private var fullMessageHead: HTTPRequestHead? = nil
+    private var buffer: ByteBuffer! = nil
+    private var maxContentLength: Int
+    private var closeOnExpectationFailed: Bool
+    private var state: AggregatorState
+    
+    public init(maxContentLength: Int, closeOnExpectationFailed: Bool = false) {
+        precondition(maxContentLength >= 0, "maxContentLength must not be negative")
+        self.maxContentLength = maxContentLength
+        self.closeOnExpectationFailed = closeOnExpectationFailed
+        self.state = .idle
+    }
+
+    public func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+        let msg = self.unwrapInboundIn(data)
+        var serverResponse: HTTPResponseHead? = nil
+
+        do {
+            switch msg {
+            case .head(let httpHead):
+                try self.state.messageHeadReceived()
+                serverResponse = self.beginAggregation(context: context, request: httpHead, message: msg)
+            case .body(var content):
+                try self.state.messageBodyReceived()
+                serverResponse = self.aggregate(context: context, content: &content, message: msg)
+            case .end(let trailingHeaders):
+                try self.state.messageEndReceived()
+                self.endAggregation(context: context, trailingHeaders: trailingHeaders)
+            }
+        } catch let error as NIOHTTPObjectAggregatorError {
+            context.fireErrorCaught(error)
+            // Won't be able to complete those
+            self.fullMessageHead = nil
+            self.buffer.clear()
+        } catch {
+            // Ignore other types of errors
+        }
```

Comment:
Though this branch is unreachable, I think it'd be better if we threw unexpected error types down the pipeline with `fireErrorCaught`.

Reply:
that's sensible, fixed